### ISSUE TITLE
Dependabot doesn't honor package.json/engine/node causing ERR_PNPM_UNSUPPORTED_ENGINE

### ISSUE THREAD

**Issue by some-user123** [dependabot/dependabot-core#7426]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package ecosystem

npm

### Package manager version

pnpm

### Language version

Node.js 18

### What you expected to see, versus what you actually saw

I'm using pnpm (a million thanks for supporting it! ✨).

I've specified to use node 18 in my package.json (engine.node):
```jsonc
{
  "engines": {
    "node": "18"
  },
  // ...
}
```

Dependabot seems not to install Node 18 but Node 16. Resulting in this error:

```
updater | 2023/06/10 18:59:28 ERROR <job_676379356>  ERR_PNPM_UNSUPPORTED_ENGINE  Unsupported environment (bad pnpm and/or Node.js version)
updater | 
updater | Your Node version is incompatible with "/home/dependabot/dependabot-updater/repo".
updater | 
updater | Expected version: 18
updater | Got: v16.20.0
updater | 
updater | This is happening because the package's manifest has an engines.node field specified.
updater | To fix this issue, install the required Node version.
```

I would expect, that Dependabot installs/uses the Node version as specified in package.json.

**Comment by matteosacchetto** [comment]
This issue is related to #6654

**Comment by deivid-rodriguez** [comment]
Indeed, this is not specific to PNPM but a current limitation that we have where we enforce the nodejs version Dependabot itself is running, failing when nodejs requirements in `package.json` file are incompatible with it.

**Comment by some-user123** [comment]
That's correct. However, I don't think, switching to Node 18 in dependabot solves the actual issue.

A good solution would be if the node version used by dependabot could be defined similarly to `actions/setup-node` (`node-version` and `node-version-file`, https://github.com/actions/setup-node/blob/main/README.md).

That would allow users to switch node versions when they want to, and not force them to keep using the same version as dependabot does.

**Comment by matteosacchetto** [comment]
I agree that being able to switch version would be the correct solution to follow

**Comment by stovmascript** [comment]
We're using Node.js 22. We restored our Dependabot compatibility by switching package managers from Yarn v4 to pnpm v9 because of this long standing issue #6107 with Yarn.

But now we find out Dependabot only supports Node.js 20, so we cannot enforce a Node.js version by using `engine-strict=true`. This needs to be fixed.

**Comment by alfaproject** [comment]
We are also hitting the same issue. We no longer support node 20, but Dependabot requires node 20...
```
updater | 2025/06/20 09:50:40 ERROR <job_1037897473> Error processing serverless (Dependabot::SharedHelpers::HelperSubprocessFailed)
2025/06/20 09:50:40 ERROR <job_1037897473> .                                        |  WARN  Unsupported engine: wanted: {"node":"^22.12"} (current: {"node":"v20.19.2","pnpm":"10.11.1"})
```

**Comment by terrymun** [comment]
This is still a recurring issue with Dependabot: it would be a huge QoL improvement if Dependabot can use the node version specified in the engines field like how `actions/setup-node` do.

Right now we are running into issues where zod cannot be auto-updated by Dependabot:

```text
2025/07/09 04:54:48 ERROR <job_1049799081> Error running package manager command: corepack pnpm update zod@3.25.74  --lockfile-only --no-save -r, Error: .                                        |  WARN  Unsupported engine: wanted: {"node":"22.17.0"} (current: {"node":"v20.19.3","pnpm":"10.12.4"})

+--------------------------------------------+
|       Dependencies failed to update        |
+------------+---------------+---------------+
| Dependency | Error Type    | Error Details |
+------------+---------------+---------------+
| zod        | unknown_error | null          |
+------------+---------------+---------------+
```

**Comment by kevbost** [comment]
Came here looking for answers to this problem and found that the [npm_and_yarn/Dockerfile](https://github.com/dependabot/dependabot-core/blob/v0.349.0/npm_and_yarn/Dockerfile#L14-L18) hardcodes the node version to v22 which is currently maintenance LTS?

I maintain a few projects that are in the process of upgrading to v24 (current LTS) but I came here expecting to find a way to enable reading from engines.node or an .nvmrc file just to find out that dependabot doesn't support that. There's an [open PR](https://github.com/dependabot/dependabot-core/pull/13576) to update to v24 but that's still not really solving a common feature that I'm sure many people come here looking for.

EDIT: PR that hard-codes node24 was merged https://github.com/dependabot/dependabot-core/pull/13576
